### PR TITLE
KAFKA-13528: KRaft RegisterBroker should validate that the cluster ID matches

### PR DESCRIPTION
**Pull request by cmccabe**
The KRaft controller should validate that the clusterID matches before allowing a broker to register in 
the cluster.

### CODE REVIEW

File: metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java [hunk 1]
```diff
@@ -84,6 +90,27 @@ public void testReplay() {
...
+        clusterControl.activate();
+        try {
+            clusterControl.registerBroker(new BrokerRegistrationRequestData().
+                    setClusterId("WIjw3grwRZmR2uOpdpVXbg").
+                    setBrokerId(0).
+                    setRack(null).
+                    setIncarnationId(Uuid.fromString("0H4fUu1xQEKXFYwB1aBjhg")),
+                123L,
+                new FeatureMapAndEpoch(new FeatureMap(Collections.emptyMap()), 456L));
+            Assertions.fail("Expected exception");
```
Comment (jsancio) [inline metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java]: Did you consider using `Assertion.assertThrows`?

**Review by jsancio — COMMENTED** [review]
What should be the broker behavior if cannot register because of an invalid cluster id? Looking at the current code it looks like the broker will continue to retry.  `InconsistentClusterIdException` is not mark as retriable so I think the broker should just shutdown, no?

Should we add an integration that shows the expected behavior on the broker side?

**Comment by cmccabe** [comment]
> What should be the broker behavior if cannot register because of an invalid cluster id? Looking at the current code it looks like the broker will continue to retry. InconsistentClusterIdException is not mark as retriable so I xxx  think the broker should just shutdown, no?

If the broker can't perform its initial registration, it will eventually shut itself down. This behavior is not specific to `InconsistentClusterIdException` -- if the broker can't register for ANY reason, it will shut down eventually. The length of time it will retry is configured by `initial.broker.registration.timeout.ms`.

I agree that maybe we could shut down more rapidly here, but it seems simpler just to rely on the standard timeout. I don't think this is a common enough scenario to need a special case optimization.

**Comment by jsancio** [comment]
Thanks for the reply @cmccabe .

Should we add an integration that shows the expected behavior on the broker side?

Can we add a description to the PR so that it is included in the commit message?

**Comment by cmccabe** [comment]
> Should we add an integration that shows the expected behavior on the broker side?

It would be a difficult integration test to write, since the existing test framework sets up the same cluster ID for everything. To be honest, I don't think writing this would be the best use of our integration-test-writing time: converting many of the existing tests that do more realistic things would have a much higher ROI. Let's file a follow-up JIRA to write an integration test like this later and try to get this in.

**Review by jsancio — APPROVED** [review]
LGTM. Thanks for the changes @cmccabe

**Pull request merged**